exonum/src/runtime/dispatcher.rs

```diff
@@ -16,143 +16,387 @@ use std::collections::HashMap;
 
     pub fn finalize(self) -> Dispatcher {
-        Dispatcher {
-            runtimes: self.runtimes,
-        }
+        Dispatcher::new(self.runtimes)
     }
 }
 
 #[derive(Default)]
 struct Dispatcher {
-    runtimes: HashMap<String, Box<dyn RuntimeEnvironment>>,
+    runtimes: HashMap<RuntimeIdentifier, Box<dyn RuntimeEnvironment>>,
+    runtime_lookup: HashMap<ServiceInstanceId, RuntimeIdentifier>,
+}
+
+impl Dispatcher {
+    pub fn new(runtimes: HashMap<RuntimeIdentifier, Box<dyn RuntimeEnvironment>>) -> Self {
+        Self {
+            runtimes,
+            runtime_lookup: Default::default(),
+        }
+    }
+
+    fn notify_service_started(&mut self, service_id: ServiceInstanceId, artifact: ArtifactSpec) {
+        let runtime_id = artifact.into();
+
+        self.runtime_lookup.insert(service_id, runtime_id);
+    }
 }
 
 impl RuntimeEnvironment for Dispatcher {
-    fn start_deploy(&self, _artifact: ArtifactSpec) -> Result<(), DeployError> {
-        Ok(())
+    fn start_deploy(&self, artifact: ArtifactSpec) -> Result<(), DeployError> {
+        let runtime_id = artifact.clone().into();
+
+        if let Some(runtime) = self.runtimes.get(&runtime_id) {
+            runtime.start_deploy(artifact)
+        } else {
+            Err(DeployError::WrongRuntime)
+        }
     }
-    fn check_deploy_status(&self, _artifact: ArtifactSpec) -> Result<DeployStatus, DeployError> {
-        Ok(DeployStatus::Deployed)
+
+    fn check_deploy_status(&self, artifact: ArtifactSpec) -> Result<DeployStatus, DeployError> {
+        let runtime_id = artifact.clone().into();
+
+        if let Some(runtime) = self.runtimes.get(&runtime_id) {
+            runtime.check_deploy_status(artifact)
+        } else {
+            Err(DeployError::WrongRuntime)
+        }
     }
 
     fn init_service(
-        &self,
-        _: &mut EnvContext,
-        _artifact: ArtifactSpec,
-        _init: &InstanceInitData,
+        &mut self,
+        ctx: &mut EnvContext,
+        artifact: ArtifactSpec,
+        init: &InstanceInitData,
     ) -> Result<(), InitError> {
-        Ok(())
+        let runtime_id = artifact.clone().into();
+
+        if let Some(runtime) = self.runtimes.get_mut(&runtime_id) {
+            let result = runtime.init_service(ctx, artifact.clone(), init);
+            if result.is_ok() {
+                self.notify_service_started(init.instance_id.clone(), artifact);
+            }
+            result
+        } else {
+            Err(InitError::WrongRuntime)
+        }
     }
 
     fn execute(
         &self,
-        _context: &mut EnvContext,
-        _dispatch: CallInfo,
-        _payload: &[u8],
+        context: &mut EnvContext,
+        call_info: CallInfo,
+        payload: &[u8],
     ) -> Result<(), ExecutionError> {
-        Ok(())
+        let runtime_id = self.runtime_lookup.get(&call_info.instance_id);
+
+        if runtime_id.is_none() {
+            return Err(ExecutionError::with_description(0x00, "Wrong runtime"));
```

Comment:
Services use errors with id 0 and it can cause problems. I think we should reserve higher error numbers for runtime errors. 
EDIT: Ok, I see todo there, so yeah we should just reserve some error numbers for system errors like (service not found, method not found, etc.)